File: clients/android/app/src/main/kotlin/my/ssdid/drive/domain/model/Invitation.kt
package my.ssdid.drive.domain.model

/**
 * Domain model representing a tenant member.
 */
data class TenantMember(
    val id: String,
    val userId: String,
    val email: String?,
    val displayName: String?,
    val role: UserRole,
    val status: MemberStatus,
    val joinedAt: String?
)

/**
 * Member status in a tenant.
 */
enum class MemberStatus {
    ACTIVE,
    PENDING,
    SUSPENDED;

    companion object {
        fun fromString(value: String): MemberStatus {
            return when (value.lowercase()) {
                "active" -> ACTIVE
                "pending" -> PENDING
                "suspended" -> SUSPENDED
                else -> ACTIVE
            }
        }
    }
}

/**
 * Domain model representing a pending invitation for the current user.
 */
data class Invitation(
    val id: String,
    val tenantId: String,
    val tenantName: String?,
    val tenantSlug: String?,
    val role: UserRole,
    val invitedBy: Inviter?,
    val invitedAt: String?
)

/**
 * Information about who sent an invitation.
 */
data class Inviter(
    val id: String?,
    val email: String?,
    val displayName: String?
) {
    /**
     * Get display text for the inviter.
     */
    fun getDisplayText(): String {
        return displayName ?: email ?: "Unknown"
    }
}

/**
 * Result of accepting an invitation.
 */
data class InvitationAccepted(
    val id: String,
    val tenantId: String,
    val role: UserRole,
    val joinedAt: String?
)

// ==================== Token Invitation (Public - for new users) ====================

/**
 * Public invitation info retrieved by token.
 * Used for invitation-only registration flow.
 */
data class TokenInvitation(
    val id: String,
    val email: String,
    val role: UserRole,
    val tenantName: String,
    val inviterName: String?,
    val message: String?,
    val expiresAt: String,
    val valid: Boolean,
    val errorReason: TokenInvitationError?
)

/**
 * Possible error reasons for invalid invitations.
 */
enum class TokenInvitationError {
    EXPIRED,
    REVOKED,
    ALREADY_USED,
    NOT_FOUND;

    companion object {
        fun fromString(value: String?): TokenInvitationError? {
            return when (value?.lowercase()) {
                "expired" -> EXPIRED
                "revoked" -> REVOKED
                "already_used" -> ALREADY_USED
                "not_found" -> NOT_FOUND
                else -> null
            }
        }
    }
}

File: clients/ios/SecureSharing/SecureSharing/Domain/Model/Invitation.swift
import Foundation

// MARK: - Token Invitation (Public - for new users)

/// Public invitation info retrieved by token.
/// Used for invitation-only registration flow.
struct TokenInvitation: Codable, Equatable {
    let id: String
    let email: String
    let role: UserRole
    let tenantName: String
    let inviterName: String?
    let message: String?
    let expiresAt: Date
    let valid: Bool
    let errorReason: TokenInvitationError?

    enum CodingKeys: String, CodingKey {
        case id
        case email
        case role
        case tenantName = "tenant_name"
        case inviterName = "inviter_name"
        case message
        case expiresAt = "expires_at"
        case valid
        case errorReason = "error_reason"
    }
}

/// Possible error reasons for invalid invitations.
enum TokenInvitationError: String, Codable, Equatable {
    case expired
    case revoked
    case alreadyUsed = "already_used"
    case notFound = "not_found"

    var displayMessage: String {
        switch self {
        case .expired:
            return "This invitation has expired"
        case .revoked:
            return "This invitation has been revoked"
        case .alreadyUsed:
            return "This invitation has already been used"
        case .notFound:
            return "Invitation not found"
        }
    }
}

/// User role in a tenant
enum UserRole: String, Codable, Equatable {
    case admin
    case member
    case viewer

    var displayName: String {
        switch self {
        case .admin:
            return "Admin"
        case .member:
            return "Member"
        case .viewer:
            return "Viewer"
        }
    }
}

// MARK: - API Response Types

/// Response from GET /invite/{token}
struct InviteInfoResponse: Codable {
    let data: TokenInvitation
}

/// Request to accept an invitation and register
struct AcceptInviteRequest: Codable {
    let displayName: String
    let password: String
    let publicKeys: AcceptInvitePublicKeys
    let encryptedMasterKey: String
    let encryptedPrivateKeys: String
    let keyDerivationSalt: String

    enum CodingKeys: String, CodingKey {
        case displayName = "display_name"
        case password
        case publicKeys = "public_keys"
        case encryptedMasterKey = "encrypted_master_key"
        case encryptedPrivateKeys = "encrypted_private_keys"
        case keyDerivationSalt = "key_derivation_salt"
    }
}

/// Public keys format for API registration
struct AcceptInvitePublicKeys: Codable {
    let kem: String          // Base64 KAZ-KEM public key
    let sign: String         // Base64 KAZ-SIGN public key
    let mlKem: String?       // Base64 ML-KEM public key
    let mlDsa: String?       // Base64 ML-DSA public key

    enum CodingKeys: String, CodingKey {
        case kem
        case sign
        case mlKem = "ml_kem"
        case mlDsa = "ml_dsa"
    }
}

/// Response from POST /invite/{token}/accept
struct AcceptInviteResponse: Codable {
    let data: AcceptInviteData
}

struct AcceptInviteData: Codable {
    let user: InviteUser
    let accessToken: String
    let refreshToken: String
    let expiresIn: Int?
    let tokenType: String?

    enum CodingKeys: String, CodingKey {
        case user
        case accessToken = "access_token"
        case refreshToken = "refresh_token"
        case expiresIn = "expires_in"
        case tokenType = "token_type"
    }
}

/// User info returned from invitation acceptance
struct InviteUser: Codable, Identifiable, Equatable {
    let id: String
    let email: String
    let displayName: String?
    let tenantId: String?
    let createdAt: Date
    let updatedAt: Date

    enum CodingKeys: String, CodingKey {
        case id
        case email
        case displayName = "display_name"
        case tenantId = "tenant_id"
        case createdAt = "created_at"
        case updatedAt = "updated_at"
    }
}

// MARK: - Authenticated Invitation (for existing users joining tenants)

/// Invitation for existing users to join additional tenants
struct TenantInvitation: Codable, Identifiable, Equatable {
    let id: String
    let tenantId: String
    let tenantName: String
    let role: UserRole
    let invitedBy: InvitedBy?
    let expiresAt: Date
    let createdAt: Date

    enum CodingKeys: String, CodingKey {
        case id
        case tenantId = "tenant_id"
        case tenantName = "tenant_name"
        case role
        case invitedBy = "invited_by"
        case expiresAt = "expires_at"
        case createdAt = "created_at"
    }
}

struct InvitedBy: Codable, Equatable {
    let id: String
    let email: String?
    let displayName: String?

    enum CodingKeys: String, CodingKey {
        case id
        case email
        case displayName = "display_name"
    }

    var name: String {
        displayName ?? email ?? "Unknown"
    }
}
